Added new compress post-processor from Vasiliy Tolstov

cbednarski opened the pull request:
Remaining work

- [ ] Full loop test with packer
- [x] Test on Windows
- [x] Try to de-dupe the test cases
- [x] Expand acceptance tests to read inside the archives
- [x] Review and cleanup

mitchellh commented on post-processor/compress/artifact.go:
```diff
@@ -5,40 +5,34 @@ import (
 )
 
-const BuilderId = "packer.post-processor.compress"
+const BuilderId = "vtolstov.compress"
```
Let's just keep the same builder ID that we had before.

mitchellh commented on post-processor/compress/benchmark.go:
```diff
@@ -0,0 +1,197 @@
+
+import (
+	"compress/flate"
+	gzip "compress/gzip"
```
Superflous "gzip" alias

mitchellh commented:
LGTM as long as its backwards compatible. 

Also, will fix #2119

cbednarski commented:
> as long as its backwards compatible.
There's no acceptance tests before / after, so I'll have to figure that out.

mitchellh commented:
Yeah, but the interface below was quite simple so it should be easy to verify.

Basic things: are the config parameters named the same, is the biulder ID the same out.

vtolstov commented:
when this can be merged ?=) thanks!

cbednarski commented:
@vtolstov I'm working on an acceptance test for this right now. Will update the PR when it's ready.

vtolstov commented:
@cbednarski thanks!

james-portman commented:
Getting in here,
have some additions I would like to add as a pull request once this is merges

cbednarski commented:
Some updates on this PR, since I've been dark for a bit:

- I added a `file` builder. This is used to create dummy test files for the compress post-processor to consume.
- I added a regression test for the version of compress on master, so we can verify that this change is backwards-compatible.
- I reworked @vtolstov's initial PR. Some of the highlights:
    - I removed `bgzp` and `gzip` in favor of `pgzip`, which seems to have the best write performance. I didn't see value in maintaining three options for the same format.
    - I simplified the configuration. Most things are automatic based on filename or environment. I did keep `compression_level` and `keep_input_artifact`.
    - I differentiate between compressors (which do compressing) and archives (which do collation). This allowed me to support @james-portman's feature pretty cleanly, and also dried up a lot of code.
- I added an acceptance test suite for various permutations of the compress feature.

Since the PR is large I recommend starting with the docs and tests.

/cc @vtolstov @james-portman @mitchellh

vtolstov commented:
@cbednarski 

* please not remove any compressors. bgzf allows seeking and best read performance.
* also not store in filename compress format. i have filename like fedora-22-x86_64 and need this name, why i need to store compress format in filename ?

james-portman commented:
@cbednarski <https://github.com/cbednarski>

Sounds good, but can I also ask not to remove compressors? In particular
normal gzip?
It is nice being able to use very small/dumb clients to unpack image files
from network sources e.g. pxe boot custom OS then ungzip and write image
straight to hard disk.
Unless the pgzip files can be unpacked by standard gzip of course..

On 18 June 2015 at 13:43, Vasiliy Tolstov <notifications@github.com> wrote:

> @cbednarski <https://github.com/cbednarski>
>
>    - please not remove any compressors. bgzf allows seeking and best read
>    performance.
>    - also not store in filename compress format. i have filename like
>    fedora-22-x86_64 and need this name, why i need to store compress format in
>    filename ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mitchellh/packer/pull/2205#issuecomment-113143264>.
>

mitchellh commented on post-processor/compress/post-processor.go:
```diff
@@ -2,97 +2,308 @@ package compress
+	KeepInputArtifact   bool   `mapstructure:"keep_input_artifact"`
+	Archive             string
+	Algorithm           string
+	UsingDefault        bool
```
Nitpick, but can you add a newline after this? I like to separate the "categories" of options, and exported/unexported usually fits the bill.

cbednarski replied:
Cool. I think I've updated this the way you suggested.

mitchellh commented on post-processor/compress/post-processor.go:
```diff
@@ -2,97 +2,308 @@ package compress
-	if err != nil {
-		return err
+
+	fmt.Printf("CompressionLevel: %d\n", p.config.CompressionLevel)
```
Debug code? :)

cbednarski replied:
Good catch, thanks :)

mitchellh commented on post-processor/compress/post-processor.go:
```diff
@@ -2,97 +2,308 @@ package compress
+		p.config.CompressionLevel = pgzip.DefaultCompression
+	}
+
+	fmt.Printf("CompressionLevel: %d\n", p.config.CompressionLevel)
```
Debug code

mitchellh commented:
@cbednarski nailed it. This looks really perfect. I'd say the only improvement that could be made (but can be done after) is to test decompression within the acceptance tests. We do that in atlas-go if you wanted to take a look. Like I said, we can do that after, but would be nice. 

This LGTM.

@vtolstov I'd be happy to take a look at re-adding bzip after, but this is still a strict improvement over our built-in one. We usually don't do such large PRs so we can break down more features into future PRs.

vtolstov commented:
@mitchellh okay
ps. bgzf  very important see benchmarks in https://github.com/vtolstov/packer-post-processor-compress

mitchellh commented:
@james-portman pgzip is just parallel gzip, it is just gzip. :)

mitchellh commented:
@vtolstov Well, like I said given that this is still strictly better than what was in Packer before, I think its fine. Nobody has really complained about compression speeds or types in Packer official issues and now we've got faster compression and multiple formats, so I'm sure bgzf will come. :)

cbednarski commented:
@james-portman pgzip, gzip, and bgfz all produce gzip artifacts. As @mitchellh mentioned, pgzip is a parallel algorithm which makes compression faster on multiple CPUs. For completeness I'm going to add a test that verifies that the archives produced by pgzip are readable using standard gzip libraries.

@vtolstov The benchmark is very useful! From the benchmark I was able to see that bgzf is best at reading, and pgzip is best at writing. Since this post-processor only writes, I thought pgzip was the best one to use. The individual cost for adding new compressors and archive formats should be pretty low now, but I did not want to make configuration super compicated.

vtolstov commented:
@cbednarski ok, bgzf may be added later, but as i say - not encode in filename archive format.
Artifact name does not need to be depend on compressor. Add additional variable for format does not make code more complicated.

cbednarski commented:
> i have filename like fedora-22-x86_64 and need this name, why i need to store compress format in filename ? ... Add additional variable for format does not make code more complicated.

@vtolstov That use-case makes sense. We can add this in a follow-up PR.

cbednarski commented:
Existing tests are passing on windows:

    cmd /C "set PACKER_ACC=1 && go test -v .\post-processor\compress"

Going to add the post-archive acceptance tests and then merge.

cbednarski commented:
There's a basic test now that reads from one of the compressed archives to verify data is intact. I'm going to add some more here but merging this for now. If there are outstanding issues or feature requests please open a new issue or PR. Thanks!
